### DISCUSSION TITLE
Support for Kodi backend?

### DISCUSSION THREAD

**Discussion by almightiest** [arabcoders/watchstate#316] [General] [answered]
Any thoughts on supporting a Kodi backend?  I love it as a local media player, especially for robust surround sound support, AV1 decoding, and upcoming DV support.  Plex/Jellyfin are less stable on Android TV compared to Kodi, but the lack of syncing between Plex while on the go and Kodi locally can be frustrating.

Example python integration using their JSON RPC API: https://forum.kodi.tv/showthread.php?tid=158373

**Comment by arabcoders** [comment]
> Any thoughts on supporting a Kodi backend?  I love it as a local media player, especially for robust surround sound support, AV1 decoding, and upcoming DV support.  Plex/Jellyfin are less stable on Android TV compared to Kodi, but the lack of syncing between Plex while on the go and Kodi locally can be frustrating.
> 
> 
> 
> Example python integration using their JSON RPC API: https://forum.kodi.tv/showthread.php?tid=158373


Hi. Unless this API is exposed over the network i.e. http i don't think it's possible to link kodi without having an agent running on the machine in which kodi is installed.

**Reply by almightiest** [reply]
I connect to it through http. Many people run a centralized headless Kodi with MySQL db and clients connect to that through http. Same goes for their json api, connection made via http. I point mine to the headless server and then all clients connect to that for updates.

**Reply by arabcoders** [reply]
> I connect to it through http. Many people run a centralized headless Kodi with MySQL db and clients connect to that through http. Same goes for their json api, connection made via http. I point mine to the headless server and then all clients connect to that for updates. 

Interesting... do you have a documentation for their API?

**Reply by almightiest** [reply]
https://kodi.wiki/view/JSON-RPC_API

https://kodi.wiki/view/JSON-RPC_API/v13

https://kodi.wiki/view/JSON-RPC_API/Examples

**Reply by arabcoders** [reply]
> https://kodi.wiki/view/JSON-RPC_API
> 
> 
> 
> https://kodi.wiki/view/JSON-RPC_API/v13
> 
> 
> 
> https://kodi.wiki/view/JSON-RPC_API/Examples


Alright Thanks i will see if it's possible to implement a kodi backend in the tool. Also Unless someone is willing to send PR with basic kodi support, I am afraid it will take me a while to investigate as i am moving houses and have many tasks queued.

**Reply by almightiest** [reply]
Totally understand.  Happy to help test and provide setup guidance if/when you get a start on it.

Do you have any kind of dev guide that be helpful to add another backend?  Would the existing architecture scale beyond 3 backends easily?

**Reply by arabcoders** [reply]
> Totally understand.  Happy to help test and provide setup guidance if/when you get a start on it.
> 
> 
> 
> Do you have any kind of dev guide that be helpful to add another backend?  Would the existing architecture scale beyond 3 backends easily?

Totally i have 2x plex,emby,jellyfin servers all synced. 

Sadly other than checking out the code in backend directory i am afraid there isn't a dev guide as the code hasn't really reached v1 yet. But mainly following the code should give you an idea of what required mostly the endpoints to extract info from the backend and a api endpoint for marking stuff as watched/unwatched. Those 3 endpoints are really required. as i have no idea how the kodi api works checking out the plex/jellyfin code should give you an idea mainly the `Import.php` and `Export.php` actions.

**Comment by alsdfjldafsjalsfjk** [comment]
Hello, i've been shopping around for a Trakt replacement (mainly scrobble feature) and i stumbled upon your service, it looks promising but i dont see Kodi support which makes me curious, is Kodi support in some kind of pipeline to be supported in the near future? 

Regards

**Reply by arabcoders** [reply, marked as answer]
Hello,

Thanks for your interest i investigated a possible implementation and it's more complicated then i expected, it's something i would like to focus on after i Stabilize the currently supported backends, once everything stabilize it's possible for me to focus on entirely different client implementation.

It's possible to get support right now via https://jellyfin.org/docs/general/clients/kodi/#jellyfin-for-kodi i think i haven't really tested.

**Reply by alsdfjldafsjalsfjk** [reply]
Thanks for replying, ill keep an eye out for future releases! Thanks for the link also!

**Comment by baggar11** [comment]
+1 for this

I know this is marked as answered but I'm not sure if this angle has been researched. It looked like mostly the Kodi API was pointed out but not really the MySQL shared library means to sync Kodi libraries across devices. I'm wondering if it would be possible to setup syncing directly to the Kodi MySQL/MariaDB library instead of trying to use the Kodi API?

Use case for me would be: I've got 3 Raspberry Pi's throughout the house hanging off TV's that share their library via MariaDB. I also have an Emby server that I use when out on vacation when I want to access my own content. I'd start a movie on vacation and finish it when I got back home. I'd always have to pull up the movie on Emby to know what time to fast forward to in Kodi.

Which begs the question. Would it be possible to get WatchStatus to get Plex/Emby/JellyFin to sync to/from a Kodi MySQL/MariaDB library?

I wrote a rough script to pull movie and tvshow resume times from a MariaDB/MySQL Kodi library.

```
#!/bin/bash
## MOVIE RESUME TIMES
# Database connection details
DB_HOST="localhost"
DB_USER="user_access"
DB_PASS="user_password"
DB_NAME="MyVideos131"

# SQL query to retrieve TV show names, episode names, and resume time in seconds where it is not null
SQL_QUERY="SELECT c00, resumeTimeInSeconds FROM movie_view WHERE resumeTimeInSeconds IS NOT NULL"

# Execute the SQL query and store the results in a variable
RESULT=$(mysql -h $DB_HOST -u $DB_USER -p$DB_PASS $DB_NAME -e "$SQL_QUERY")

# Print each show name, episode name, and resume time on a new line
echo "$RESULT"
```
```
#!/bin/bash
## TVSHOW RESUME TIMES
# Database connection details
DB_HOST="localhost"
DB_USER="user_access"
DB_PASS="user_password"
DB_NAME="MyVideos131"

# SQL query to retrieve TV show names, episode names, and resume time in seconds where it is not null
SQL_QUERY="SELECT strTitle, strFileName, resumeTimeInSeconds FROM episode_view WHERE resumeTimeInSeconds IS NOT NULL"

# Execute the SQL query and store the results in a variable
RESULT=$(mysql -h $DB_HOST -u $DB_USER -p$DB_PASS $DB_NAME -e "$SQL_QUERY")

# Print each show name, episode name, and resume time on a new line
echo "$RESULT"
```

**Reply by arabcoders** [reply]
Hello, Unfortunately, This is not feasible for us we only work with APIs, as WatchState can be hosted remotely, and this require WatchState to have direct access to the mysql server which isn't really good idea, unless someone come up with standard API i am afraid this out of my hands. However as workaround it was suggested that https://jellyfin.org/docs/general/clients/kodi/#jellyfin-for-kodi can be used for this case.

**Reply by almightiest** [reply]
@arabcoders this is the standard way to integrate with Kodi via API: https://kodi.wiki/view/JSON-RPC_API/v13

**Reply by arabcoders** [reply]
I remember checking out the API, there was some issues as how we handle data locally that isn't compatible with the RPC API at that time i may revisit it at some point in the future.

**Reply by baggar11** [reply]
What about using one of the APIs out there to connect to the MySQL/MariadB Kodi library for querying? It would be pretty easy to setup a read only dB user to remotely query the dB.

I had AI translate my script above using the python connector for MySQL just to show an example. Not sure if it would work or not. Thanks for taking a look!

```
import mysql.connector

# Establish a connection to your database
cnx = mysql.connector.connect(
    user='your_username',
    password='your_password',
    host='localhost',  # The address of your MySQL server
    database='MyVideosVER'
)

cursor = cnx.cursor()

# Execute the SQL query
query = ("SELECT c00 FROM movie_view WHERE resumeTimeInSeconds IS NOT NULL")
cursor.execute(query)

# Fetch all rows from the result set
rows = cursor.fetchall()

for row in rows:
    print(row[0])

# Close the connection to the database
cnx.close()
```

**Reply by arabcoders** [reply]
@baggar11 it's not possible to do what you are asking, i am not about to create entire kodi client. closest thing is probably going to be the json rpc-api
